scripts/fix-daily-branch

```diff
@@ -0,0 +1,85 @@
+#!/bin/bash
+
+# Fix daily build branches so daily build recipes by dropping all 'cherry-pick'
+# quilt patches applied to a given ubuntu/* release branch.
+#  Any *cpick* files in debian/patches are git revert'd in reverse order
+# so that the upstream commits the *cpick* file represent are able to apply.
+set -e
+
+error() { echo "$@" 1>&2; }
+fail() { [ $# -eq 0 ] || error "$@"; exit 1; }
+
+gitcmd() {
+    git "$@" || fail "Failed running git $@"
+}
+
+Usage() {
+cat <<EOF
+${0##*/}
+
+Create a daily_branch which reverts any debian/patches/*cpick* commits from
+source_branch.
+
+options:
+   -h|--help                   show this message.
+   --source-branch          The local branch name from which to revert all
+                            debian/patches/*cpick* files. Default: ubuntu/devel.
+   --daily-branch           The local daily branch name to create which reverts
+                            debian/patches/*cpick* files. Default:
+                            ubuntu/daily/devel
+EOF
+}
+
+drop_cherry_picks_from_daily_branch() {
+    local source_branch="${1}" daily_branch="${2}" daily_remote="${3:-origin}"
+    gitcmd fetch ${daily_remote}
+    echo "Backing out all debian/patches/*cpick* from ${daily_branch} branch"
+    if git branch | grep -q ${daily_branch}; then
+        gitcmd checkout ${daily_branch}
+    else
+        # Create local daily branch from current
+        gitcmd checkout ${source_branch} -b ${daily_branch}
+    fi
+    gitcmd reset --hard ${source_branch}
+    gitcmd log --oneline -- debian/patches/*cpick* | cut -d" " -f1 | xargs git revert
+    # switch back to current development branch
+    gitcmd checkout ${source_branch}
+    cat <<EOF
+To fix daily recipe builds perform the following:
+
+  # Verify local changes in $daily_branch versus $source_branch
+  git diff $daily_branch
+  git push $daily_remote $daily_branch
+EOF
+}
+
+main() {
+    local short_opts="hd:s:r:"
+    local long_opts="help,daily-branch:,source_branch:,remote:"
+    local getopt_out=""
+    getopt_out=$(getopt --name "${0##*/}" \
+        --options "${short_opts}" --long "${long_opts}" -- "$@") &&
+        eval set -- "${getopt_out}" ||
+        { Usage; return; }
+
+    local cur="" next="" daily_branch="ubuntu/daily/devel"
+    local source_branch="ubuntu/devel" remote="origin"
```

Comment:
The cherry-pick workflow was originally designed for applying hotfixes to stable releases, not to devel. As such, I don't think there is a clear default for these options, it will be situation dependent. I'm wary of just applying our most recently experienced situation as the default.

I think `origin` as the remote is also a misleading default; generally people will want to use `upstream`, and it will be confusing that the default isn't that. That said, I agree that we don't want the default to be `upstream`, because that makes it easy for accidents to happen. This is another case where I'm not sure that a default serves us very well; it's better to ask people to be explicit about what they are asking the script to do.

I do understand the desire for defaults; we don't want people to have to type out "ubuntu/daily/devel" and "ubuntu/devel" in the simplest case, because there's room for typos there. I wonder if this is actually indicating that the script should be doing a little bit more: instead of the input being (daily branch, source branch, remote), what if the input were (release, remote), and the script constructed the branch names itself? Then folks need only type out the part that's actually different, and it's more reasonable to not have a default.

Reply:
Good suggestion, I've changed the code to require 2 positionals fix-release-branch <release> <remote>